moveit_ros/perception/pointcloud_octomap_updater/src/pointcloud_octomap_updater.cpp

```diff
@@ -348,16 +358,85 @@ void PointCloudOctomapUpdater::cloudMsgCallback(const sensor_msgs::PointCloud2::
   catch (...)
   {
     ROS_ERROR_NAMED(LOGNAME, "Internal error while updating octree");
+    success = false;
   }
   tree_->unlockWrite();
-  ROS_DEBUG_NAMED(LOGNAME, "Processed point cloud in %lf ms", (ros::WallTime::now() - start).toSec() * 1000.0);
-  tree_->triggerUpdateCallback();
 
   if (filtered_cloud)
   {
     sensor_msgs::PointCloud2Modifier pcd_modifier(*filtered_cloud);
     pcd_modifier.resize(filtered_cloud_size);
     filtered_cloud_publisher_.publish(*filtered_cloud);
   }
+
+  ROS_DEBUG_NAMED(LOGNAME, "Processed point cloud in %lf ms", (ros::WallTime::now() - start).toSec() * 1000.0);
+  tree_->triggerUpdateCallback();
+
+  return success;
+}
+
+bool PointCloudOctomapUpdater::processCloud(const sensor_msgs::PointCloud2::ConstPtr& cloud_msg)
+{
+  if (monitor_->getMapFrame().empty())
+    monitor_->setMapFrame(cloud_msg->header.frame_id);
+
+  /* get transform for cloud into map frame */
+  Eigen::Isometry3d sensor_origin_eigen;
+
+  if (monitor_->getMapFrame() == cloud_msg->header.frame_id)
+    sensor_origin_eigen = Eigen::Isometry3d::Identity();
+  else
+  {
+    if (tf_buffer_)
+    {
+      try
+      {
+        geometry_msgs::TransformStamped sensor_origin =
+            tf_buffer_->lookupTransform(monitor_->getMapFrame(), cloud_msg->header.frame_id, cloud_msg->header.stamp);
+        sensor_origin_eigen = tf2::transformToEigen(sensor_origin);
+      }
+      catch (tf2::TransformException& ex)
+      {
+        ROS_ERROR_STREAM_NAMED(LOGNAME, "Transform error of sensor data: " << ex.what() << "; quitting callback");
+        return false;
+      }
+    }
+    else
+      return false;
+  }
+
+  if (!updateTransformCache(cloud_msg->header.frame_id, cloud_msg->header.stamp))
+  {
+    ROS_ERROR_THROTTLE_NAMED(1, LOGNAME, "Transform cache was not updated. Self-filtering may fail.");
+    return false;
+  }
+
+  return processCloud(cloud_msg, sensor_origin_eigen, update_method_);
 }
+
+void PointCloudOctomapUpdater::cloudMsgCallback(const sensor_msgs::PointCloud2::ConstPtr& cloud_msg)
+{
+  ROS_DEBUG_NAMED(LOGNAME, "Received a new point cloud message");
+  std::lock_guard<std::mutex> lock(update_mutex_);
```

Comment:
Wouldn't it make sense to only lock `processCloud()` to allow thread-safe usage of the C++ API?